gitcache/main.go

```diff
@@ -4,139 +4,190 @@ import (
 	"fmt"
 	"io/ioutil"
 	"log"
-	"net/url"
 	"os"
 	"path"
-	"strings"
 	"time"
 
 	"github.com/go-git/go-git/v5"
 	"github.com/mholt/archiver"
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
 )
 
 func main() {
-	blob := os.Getenv("BLOB_URL")
+	const gittar string = "gitfolder.tar.gz"
+	var gitRepo *git.Repository
 
-	if blob == "" {
-		log.Panic("BLOB_URL env is not set.")
-	}
+	logLevel := zap.LevelFlag("verbosity", zap.InfoLevel, "override the default log level")
 
-	repo := &RepoURL{}
-	err := repo.Set(os.Args[1])
+	cfg := zap.NewProductionConfig()
+	cfg.Level.SetLevel(*logLevel)
+	logger, err := cfg.Build()
 	if err != nil {
 		log.Panic(err)
 	}
+	//nolint
+	defer logger.Sync() // flushes buffer, if any
+	sugar := logger.Sugar()
+	blob := os.Getenv("BLOB_URL")
+	sugar.Info("BLOB_URL ", blob)
 
-	cwd, err := os.Getwd()
-	if err != nil {
-		log.Panic(err)
+	if blob == "" {
+		sugar.Panic("BLOB_URL env is not set.")
 	}
 
-	dir, err := ioutil.TempDir(path, repo.Owner+repo.Owner)
-	if err != nil {
-		log.Panic(err)
+	// opening the blob
+	bucket, e := NewBucket(blob)
+	if e != nil {
+		sugar.Panic(e)
 	}
 
-	tarDir, err := ioutil.TempDir(cwd, repo.Owner+repo.Owner+"tar")
+	sugar.Debugf("The bucket was opened %s", blob)
+
+	repo := RepoURL{}
+	err = repo.Set(os.Args[1])
 	if err != nil {
-		log.Panic(err)
+		sugar.Panic(err)
 	}
 
-	defer os.RemoveAll(dir)
-	defer os.RemoveAll(tarDir)
+	sugar.Debugf("Fetching %+v", repo)
 
-	// Clone the given repository to the given directory
-	r, err := git.PlainClone(dir, false, &git.CloneOptions{
-		URL:   fmt.Sprintf("http://%s/%s/%s", repo.Host, repo.Owner, repo.Repo),
-		Depth: 1, // Just fetch the last commit
-	})
+	// gets all the path configuration.
+	storage, err := getStoragePath(repo)
 	if err != nil {
-		log.Panic(err)
+		sugar.Panic(err)
+	}
+
+	defer os.RemoveAll(storage.GitDir)
+	defer os.RemoveAll(storage.GitTarDir)
+	defer os.RemoveAll(storage.BlobArchiveDir)
```

Comment:
It would be better to make a method so that we can do `defer storage.Cleanup()` rather than expecting the caller to know which ones to cleanup.

Reply:
Good suggestion! Thanks, implemented it.